packages/cohere/src/cohere-chat-language-model.ts

```diff
@@ -304,3 +367,99 @@ const cohereChatChunkSchema = z.discriminatedUnion('event_type', [
     }),
   }),
 ]);
+
+// For reference: https://docs.cohere.com/docs/parameter-types-in-tool-use
+
+function prepareToolsAndToolChoice(
+  mode: Parameters<LanguageModelV1['doGenerate']>[0]['mode'] & {
+    type: 'regular';
+  },
+) {
+  const tools = mode.tools?.length ? mode.tools : undefined;
+
+  if (tools == null) {
+    return { tools: undefined };
+  }
+
+  const mappedTools = tools.map(tool => {
+    const { properties, required } = tool.parameters;
+
+    const parameterDefinitions: any = {};
+
+    if (properties) {
+      for (const [key, value] of Object.entries(properties)) {
+        if (typeof value === 'object' && value !== null) {
+          const { type: JSONType, description } = value;
+
+          let type: 'str' | 'float' | 'int' | 'bool';
+
+          if (typeof JSONType === 'string') {
+            switch (JSONType) {
+              case 'string':
+                type = 'str';
+                break;
+              case 'number':
+                type = 'float';
+                break;
+              case 'integer':
+                type = 'int';
+                break;
+              case 'boolean':
+                type = 'bool';
+                break;
+              default:
+                throw new Error(`Unsupported JSON type: ${JSONType}`);
+            }
+          } else if (Array.isArray(JSONType)) {
+            throw new Error('Array types are not supported');
+          } else {
+            throw new Error('Unsupported type');
+          }
+
+          parameterDefinitions[key] = {
+            required: required ? required.includes(key) : false,
+            type,
+            description,
+          };
+        }
+      }
+    }
+
+    return {
+      name: tool.name,
+      description: tool.description,
+      parameterDefinitions,
+    };
+  });
+
+  const toolChoice = mode.toolChoice;
+
+  if (toolChoice == null) {
+    return { tools: mappedTools, force_single_step: false };
+  }
+
+  const type = toolChoice.type;
+
+  switch (type) {
+    case 'auto':
+      return { tools: mappedTools, force_single_step: false };
+    case 'required':
+      return { tools: mappedTools, force_single_step: true };
+
+    // cohere does not support 'none' tool choice, so we remove the tools:
+    case 'none':
+      return { tools: undefined, force_single_step: false };
+
+    // cohere does not support tool mode directly,
+    // so we filter the tools and force the tool choice through 'any'
+    case 'tool':
+      return {
+        tools: mappedTools.filter(tool => tool.name === toolChoice.toolName),
+        force_single_step: true,
```

Comment:
what does this do?

Reply:
`force_single_step` is cohere's way of forcing the model to strictly generate a tool call during a generation

link to docs: https://docs.cohere.com#when-should-i-use-single-step-tool-use